### PR TITLE
Initial implementation of unformed state for local variables.

### PR DESCRIPTION
**Pull request by SlaterLatiao**
Allows unformed state for local variables. Reports a run-time error when an unformed local variable is used.
- Added declaration without initialization for local variables in the parser.
- Made the init expression of VariableDefinition optional.
- Expanded (alive, dead) to (uninitialized, alive, dead) in the Heap.

### CODE REVIEW

File: explorer/syntax/parser.ypp [hunk 1]
```diff
@@ -756,6 +756,13 @@ clause_list:
 statement:
   statement_expression EQUAL expression SEMICOLON
     { $$ = arena->New<Assign>(context.source_loc(), $1, $3); }
+
```
[the remark this answers is not in the document]
Comment (SlaterLatiao) [reply explorer/syntax/parser.ypp]: Fixed.

File: explorer/interpreter/heap.h [hunk 2]
```diff
@@ -19,6 +19,12 @@ namespace Carbon {
 // A Heap represents the abstract machine's dynamically allocated memory.
 class Heap : public HeapAllocationInterface {
  public:
+  enum ValueState {
```
[the remark this answers is not in the document]
Comment (SlaterLatiao) [reply explorer/interpreter/heap.h]: Fixed.

File: explorer/ast/statement.h [hunk 3]
```diff
@@ -131,17 +131,26 @@ class VariableDefinition : public Statement {
...
-
+  auto has_init() const -> bool { return init_.has_value(); }
+  auto init() const -> const Expression& {
+    CARBON_CHECK(has_init());
+    return **init_;
+  }
+  auto init() -> Expression& {
+    CARBON_CHECK(has_init());
+    return **init_;
+  }
```
[the remark this answers is not in the document]
Comment (SlaterLatiao) [reply explorer/ast/statement.h]: Done.

File: explorer/interpreter/interpreter.cpp [hunk 4]
```diff
@@ -251,6 +251,18 @@ auto PatternMatch(Nonnull<const Value*> p, Nonnull<const Value*> v,
           }  // for
           return true;
         }
+        case Value::Kind::UninitializedValue: {
+          const auto& p_tup = cast<TupleValue>(*p);
+          for (size_t i = 0; i < p_tup.elements().size(); ++i) {
```
[the remark this answers is not in the document]
Comment (SlaterLatiao) [reply explorer/interpreter/interpreter.cpp]: Used range-based for loop instead.

**Review by jonmeow — APPROVED** [review]

**Pull request merged**